packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraProperties.java

```diff
@@ -0,0 +1,166 @@
+import android.hardware.camera2.CameraManager;
+import android.os.Build.VERSION_CODES;
+import android.util.Range;
+import android.util.Rational;
+import android.util.Size;
+import androidx.annotation.RequiresApi;
+
+/** An interface allowing access to the different characteristics of the device's camera. */
+public interface CameraProperties {
+  String getCameraName();
+
+  Range<Integer>[] getControlAutoExposureAvailableTargetFpsRanges();
+
+  Range<Integer> getControlAutoExposureCompensationRange();
+
+  double getControlAutoExposureCompensationStep();
+
+  int[] getControlAutoFocusAvailableModes();
+
+  Integer getControlMaxRegionsAutoExposure();
+
+  Integer getControlMaxRegionsAutoFocus();
+
+  int[] getDistortionCorrectionAvailableModes();
+
+  Boolean getFlashInfoAvailable();
+
+  int getLensFacing();
+
+  Float getLensInfoMinimumFocusDistance();
+
+  Float getScalerAvailableMaxDigitalZoom();
+
+  Rect getSensorInfoActiveArraySize();
+
+  Size getSensorInfoPixelArraySize();
+
+  Rect getSensorInfoPreCorrectionActiveArraySize();
+
+  int getSensorOrientation();
+
+  int getHardwareLevel();
+
+  int[] getAvailableNoiseReductionModes();
+}
+
+/**
+ * Implementation of the @see CameraProperties interface using the @see
+ * android.hardware.camera2.CameraCharacteristics class to access the different characteristics.
+ */
+class CameraPropertiesImpl implements CameraProperties {
+  private final CameraCharacteristics cameraCharacteristics;
+  private final String cameraName;
+
+  public CameraPropertiesImpl(String cameraName, CameraManager cameraManager)
+      throws CameraAccessException {
+    this.cameraName = cameraName;
+    this.cameraCharacteristics = cameraManager.getCameraCharacteristics(cameraName);
+  }
+
+  @Override
+  public String getCameraName() {
+    return cameraName;
+  }
+
+  @Override
+  public Range<Integer>[] getControlAutoExposureAvailableTargetFpsRanges() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_AE_AVAILABLE_TARGET_FPS_RANGES);
+  }
+
+  @Override
+  public Range<Integer> getControlAutoExposureCompensationRange() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_AE_COMPENSATION_RANGE);
+  }
+
+  @Override
+  public double getControlAutoExposureCompensationStep() {
+    Rational rational =
+        cameraCharacteristics.get(CameraCharacteristics.CONTROL_AE_COMPENSATION_STEP);
+
+    return rational == null ? 0.0 : rational.doubleValue();
+  }
+
+  @Override
+  public int[] getControlAutoFocusAvailableModes() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_AF_AVAILABLE_MODES);
+  }
+
+  @Override
+  public Integer getControlMaxRegionsAutoExposure() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_MAX_REGIONS_AE);
+  }
+
+  @Override
+  public Integer getControlMaxRegionsAutoFocus() {
+    return cameraCharacteristics.get(CameraCharacteristics.CONTROL_MAX_REGIONS_AF);
+  }
+
+  @RequiresApi(api = VERSION_CODES.P)
```

Comment:
Maybe this is a Java noob question, but shouldn't this be on the interface, not the implementation? It's not clear to me what it means for an interface with no API requirement to be implemented by something that does have one.